go/upgrade/upgrade.go

```diff
@@ -0,0 +1,253 @@
+// Package upgrade implements the node upgrade backend.
+//
+// After submitting an upgrade descriptor, the old node may continue
+// running or be restarted up to the point when the consensus layer reaches
+// the upgrade epoch. The new node may not be started until the old node has
+// reached the upgrade epoch.
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	"github.com/oasislabs/oasis-core/go/common/crypto/hash"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	"github.com/oasislabs/oasis-core/go/common/persistent"
+	"github.com/oasislabs/oasis-core/go/common/version"
+	epochtime "github.com/oasislabs/oasis-core/go/epochtime/api"
+	"github.com/oasislabs/oasis-core/go/upgrade/api"
+	"github.com/oasislabs/oasis-core/go/upgrade/migrations"
+)
+
+var (
+	_ api.Backend = (*upgradeManager)(nil)
+
+	metadataStoreKey = []byte("descriptor")
+
+	thisVersion = makeVersionString()
+)
+
+func hashSelf() (*hash.Hash, error) {
+	path, err := os.Executable()
+	if err != nil {
+		return nil, err
+	}
+
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var hash hash.Hash
+	hash.FromBytes(contents)
+	return &hash, nil
+}
+
+func makeVersionString() string {
+	return version.SoftwareVersion
+}
+
+type upgradeManager struct {
+	store   *persistent.ServiceStore
+	pending *api.PendingUpgrade
+	lock    sync.Mutex
+
+	ctx     *migrations.Context
+	handler migrations.Handler
+
+	logger *logging.Logger
+}
+
+func (u *upgradeManager) SubmitDescriptor(ctx context.Context, descriptor *api.Descriptor) error {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	u.pending = &api.PendingUpgrade{
```

Comment:
So submitting a new descriptor just overwrites the old one? Should we perhaps have an explicit "cancel pending upgrade" action?